### PR TITLE
Randomization

### LINKED ISSUE #163: Randomization
- [ ] #164 
- [ ] #165

### PR DESCRIPTION
**Pull request by ZachCutler04**
### Does this PR close any open issues?
Closes #163 

### Give a longer description of what this PR addresses and why it's needed
Adds randomization on both a component and a sequence level. Allows for choice between pure random and latin squares solution, latin square solution stores latin squares in the firebase and pulls from there. 

### Are there any additional TODOs before this PR is ready to go?
TODOs:
- [ ] Update relevant documentation
- [ ] ...

### CODE REVIEW

File: public/mvnv/config-html-MVNV.hjson [hunk 1]
```diff
@@ -628,21 +628,26 @@
...
+            trials8
+            trials9
+            trials10
+        ]
+        [   trials11
+            trials12
+            trials13
+            trials14
+            trials15
+        ]
```
Comment (hilsonshrestha) [inline public/mvnv/config-html-MVNV.hjson:649]: Should the nested sequence have their own containers?

Comment (ZachCutler04) [reply public/mvnv/config-html-MVNV.hjson:649]: Ya, Alex also mentioned this, I need a better way to define randomness within the overall sequence. Open to suggestions :D but probably transforming the sequence list into an object to allow for more specificity

File: src/components/Shell.tsx [hunk 2]
```diff
@@ -52,13 +54,51 @@ async function fetchStudyConfig(configLocation: string, configKey: string) {
...
-  globalConfig: GlobalConfig;
-};
+function orderObjectToList(
+  order: OrderObject, 
+  pathsFromFirebase: {
+    path: string;
+    order: string[];
+  }[] | undefined,
+ path: string
+ ) : string[] {
```
Comment (JackWilb) [inline src/components/Shell.tsx:64]: Implement this in 2 functions, 1 to be called and 1 that is called recursively

**Review by JackWilb — COMMENTED** [review]
Questions:
1) Does this generate blank entries for a participant if they are not shown a component?
2) Verify that restoring works
3) Mutually exclusive component keys

File: src/storage/init.ts [hunk 3]
```diff
@@ -240,6 +260,37 @@ export async function loadProvenance(
...
+        if(!currArr) {
+          const newSquare: string[][] = latinSquare<string>(data['options'].sort(() => 0.5 - Math.random()), true);
+
+          returnRefs.push({path,  order: newSquare[0] });
+          batch.set(randomRef, { perms: newSquare.slice(1).map((perm: string[]) => Object.assign({}, perm)), options: data.options });
+        }
+        else {
+          returnRefs.push({path, order: Object.values(currArr) });
+          batch.set(randomRef, { perms: data.perms, options: data.options });
+        }
```
[the remark this answers is not in the document]
Comment (JackWilb) [reply src/storage/init.ts:284]: Issue

File: src/storage/init.ts [hunk 4]
```diff
@@ -262,11 +313,79 @@ async function saveProvenanceNode(
...
+    const returnRefs = await getRandomOrders(store, paths, batch, studyId, config);
+
+    batch.commit();
+
+    return Promise.resolve(returnRefs);
+  }
+
+  paths.forEach((path) => {
+    const randObj: { perms?: Record<number, string>[], options?: string[] } = {};
+
```
[the remark this answers is not in the document]
Comment (JackWilb) [reply src/storage/init.ts:346]: Issue

File: src/parser/types.ts [hunk 5]
```diff
@@ -247,32 +249,39 @@ export interface QuestionnaireComponent extends BaseIndividualComponent {
...
-export type StudyComponent = IndividualComponent | ContainerComponent;
-
 export interface StudyComponents {
-  [key: string]: StudyComponent;
+  [key: string]: IndividualComponent;
 }
 
+export interface InheritedStudyComponents {
+  [key: string]: InheritedComponent;
+}
```
[the remark this answers is not in the document]
Comment (ZachCutler04) [reply src/parser/types.ts:266]: done

File: src/parser/types.ts [hunk 6]
```diff
@@ -247,32 +249,39 @@ export interface QuestionnaireComponent extends BaseIndividualComponent {
...
 /**
  * The StudyConfig interface is used to define the properties of a study configuration. These are the hjson files that live in the public folder. In our repo, one example of this would be public/cleveland/config-cleveland.hjson. 
  */
 export interface StudyConfig {
   $schema: string;
   studyMetadata: StudyMetadata;
   uiConfig: UIConfig;
-  components: StudyComponents;
+  baseComponents?: StudyComponents;
+  components: InheritedStudyComponents
```
[the remark this answers is not in the document]
Comment (ZachCutler04) [reply src/parser/types.ts:280]: done

File: src/components/response/TextAreaInput.tsx [hunk 7]
```diff
@@ -20,11 +20,13 @@ export default function TextAreaInput({
       <Textarea
         disabled={disabled}
         placeholder={placeholder}
-        label={prompt}
+        label={prompt}
```
[the remark this answers is not in the document]
Comment (ZachCutler04) [reply src/components/response/TextAreaInput.tsx:23]: done

File: src/storage/init.ts [hunk 8]
```diff
@@ -25,6 +29,19 @@ function serverTimestamp() {
   return new Date();
 }
 
+function createRandomOrders(order: OrderObject, paths: string[], path: string, index = 0) {
+    const newPath = path.length > 0 ? path + '-' + index : 'root';
```
[the remark this answers is not in the document]
Comment (ZachCutler04) [reply src/storage/init.ts:33]: done

File: src/storage/init.ts [hunk 9]
```diff
@@ -262,11 +313,79 @@ async function saveProvenanceNode(
...
+  config: StudyConfig,
+  studyId: string,
+) : Promise<{path: string, order: string[]}[]> {
+  const batch = writeBatch(store);
+
+  const studiesRef = doc(store, ...[STUDIES, studyId, 'versions', config.studyMetadata.version]);
+
+  const paths: string[] = [];
+
+  createRandomOrders(config.sequence, paths, '');
```
[the remark this answers is not in the document]
Comment (ZachCutler04) [reply src/storage/init.ts:332]: done

File: src/components/Shell.tsx (hunk 2 above)
Comment (ZachCutler04) [reply src/components/Shell.tsx:64]: done

File: src/controllers/ComponentController.tsx [hunk 10]
```diff
@@ -19,18 +20,15 @@ export default function ComponentController() {
...
-  // Get the id for the current trial (if there is one)
-  const trialId = useCurrentTrial();
-  const trialConfig = trialId !== null ? (stepConfig as ContainerComponent).components[trialId] : null;
   const trialProvenance = createTrialProvenance();
   
   // If we have a trial, use that config to render the right component else use the step
   const status = useComponentStatus();
 
-  const currentConfig = trialConfig !== null ? trialConfig : stepConfig;
+  const currentConfig = isPartialComponent(stepConfig) && studyConfig.baseComponents ? merge({}, studyConfig.baseComponents?.[stepConfig.baseComponent], stepConfig) as IndividualComponent : stepConfig as IndividualComponent;
```
[the remark this answers is not in the document]
Comment (ZachCutler04) [reply src/controllers/ComponentController.tsx:28]: Created here https://github.com/revisit-studies/study/issues/175

File: src/storage/init.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ZachCutler04) [reply src/storage/init.ts:284]: https://github.com/revisit-studies/study/issues/176 issue here

File: src/storage/init.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ZachCutler04) [reply src/storage/init.ts:346]: created here https://github.com/revisit-studies/study/issues/176

File: src/storage/init.ts [hunk 11]
```diff
@@ -262,11 +313,79 @@ async function saveProvenanceNode(
...
 }
 
+async function saveStudyConfig(
+  store: Firestore,
+  config: StudyConfig,
+  studyId: string,
+) : Promise<{path: string, order: string[]}[]> {
+  const batch = writeBatch(store);
+
+  const studiesRef = doc(store, ...[STUDIES, studyId, 'versions', config.studyMetadata.version]);
```
[the remark this answers is not in the document]
Comment (ZachCutler04) [reply src/storage/init.ts:328]: done

File: src/storage/downloadTidy.ts [hunk 12]
```diff
@@ -82,59 +77,15 @@ export async function downloadTidy(
...
-          endTime,
-          duration,
-        };
-
-        trs.push(tr);
-      });
-    }
-  });
-
   return trs;
```
[the remark this answers is not in the document]
Comment (ZachCutler04) [reply src/storage/downloadTidy.ts:90]: fixed it

File: src/components/interface/AppAside.tsx [hunk 13]
```diff
@@ -1,106 +1,28 @@
...
-                      </Text>
-                    );
-                  })}
-              </Paper>
-            </Paper>
-
-            <Space h="md" />
-          </React.Fragment>
-        ))}
-      </Aside.Section>
```
[the remark this answers is not in the document]
Comment (ZachCutler04) [reply src/components/interface/AppAside.tsx:103]: Added back the aside

File: src/components/interface/AppNavBar.tsx [hunk 14]
```diff
@@ -15,16 +17,15 @@ export default function AppNavBar() {
   // Get the config for the current step
   const studyConfig = useStudyConfig();
   const step = useCurrentStep();
-  const currentStepConfig = studyConfig.components[step];
+  const stepConfig = studyConfig.components[step];
 
-  const trialId = useCurrentTrial();
+  const currentConfig = isPartialComponent(stepConfig) && studyConfig.baseComponents ? merge({}, studyConfig.baseComponents?.[stepConfig.baseComponent], stepConfig) as IndividualComponent : stepConfig as IndividualComponent;
```
Comment (JackWilb) [inline src/components/interface/AppNavBar.tsx]: This and some other spots could lead to some issues with the config. I think we need to fail more gracefully. Let's open an issue for that

**Review by JackWilb — APPROVED** [review]
Looks good. Let's make that 1 issue and remove the console.logs, though

File: src/components/interface/AppNavBar.tsx (hunk 14 above)
Comment (JackWilb) [reply src/components/interface/AppNavBar.tsx]: #178

**Pull request merged**